File: python/big_o_analysis/ex2_space_complexity.py
"""
Exercise 2: Space Complexity Analysis (Easy)

Analyze the space complexity of each function below. Space complexity measures
the additional memory an algorithm uses beyond the input.

Remember:
- Input size doesn't count toward space complexity
- Count auxiliary space (variables, data structures, recursion stack)
- O(1) = constant space, O(n) = linear space, etc.

Difficulty: Easy
Topics: Space Complexity, Memory Analysis
"""

from typing import List, Dict


def snippet1(arr: List[int]) -> int:
    """Snippet 1: Sum array elements"""
    total = 0
    for num in arr:
        total += num
    return total


def get_space_complexity_snippet1() -> str:
    """
    Return the space complexity of snippet1.

    Hint: How many extra variables do we use regardless of input size?
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def snippet2(arr: List[int]) -> List[int]:
    """Snippet 2: Create a copy of the list"""
    copy = []
    for num in arr:
        copy.append(num)
    return copy


def get_space_complexity_snippet2() -> str:
    """
    Return the space complexity of snippet2.

    Hint: How does the size of 'copy' relate to the input size?
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def snippet3(n: int) -> int:
    """Snippet 3: Recursive factorial"""
    if n <= 1:
        return 1
    return n * snippet3(n - 1)


def get_space_complexity_snippet3() -> str:
    """
    Return the space complexity of snippet3.

    Hint: Consider the call stack. How many recursive calls are made?
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def snippet4(n: int) -> List[List[int]]:
    """Snippet 4: Create 2D matrix"""
    matrix = []
    for i in range(n):
        row = []
        for j in range(n):
            row.append(i + j)
        matrix.append(row)
    return matrix


def get_space_complexity_snippet4() -> str:
    """
    Return the space complexity of snippet4.

    Hint: How many elements are in an n×n matrix?
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def snippet5(arr: List[int], threshold: int) -> List[int]:
    """Snippet 5: Filter array into list"""
    result = []
    for num in arr:
        if num > threshold:
            result.append(num)
    return result


def get_space_complexity_snippet5() -> str:
    """
    Return the space complexity of snippet5.

    Hint: In the worst case, how many elements might be in the result list?
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


def snippet6(n: int, memo: Dict[int, int] = None) -> int:
    """Snippet 6: Fibonacci with memoization"""
    if memo is None:
        memo = {}
    if n <= 1:
        return n
    if n in memo:
        return memo[n]

    memo[n] = snippet6(n - 1, memo) + snippet6(n - 2, memo)
    return memo[n]


def get_space_complexity_snippet6() -> str:
    """
    Return the space complexity of snippet6.

    Hint: Consider both the memo dict size and the recursion stack depth.
    """
    # TODO: Implement this
    raise NotImplementedError("Not implemented yet")


if __name__ == "__main__":
    print("Big O Analysis - Exercise 2: Space Complexity")
    print("Implement get_space_complexity functions\n")

    try:
        print(f"Snippet 1 (sum array): {get_space_complexity_snippet1()}")
        print(f"Snippet 2 (copy array): {get_space_complexity_snippet2()}")
        print(f"Snippet 3 (recursive factorial): {get_space_complexity_snippet3()}")
        print(f"Snippet 4 (2D matrix): {get_space_complexity_snippet4()}")
        print(f"Snippet 5 (filter array): {get_space_complexity_snippet5()}")
        print(f"Snippet 6 (fibonacci memoization): {get_space_complexity_snippet6()}")
    except NotImplementedError:
        print("Not yet implemented!")

File: java/src/main/java/com/learning/bigoanalysis/ex2SpaceComplexity.java
package com.learning.bigoanalysis;

import java.util.*;

/**
 * Exercise 2: Space Complexity Analysis (Easy)
 *
 * Analyze the space complexity of each method below. Space complexity measures
 * the additional memory an algorithm uses beyond the input.
 *
 * Remember:
 * - Input size doesn't count toward space complexity
 * - Count auxiliary space (variables, data structures, recursion stack)
 * - O(1) = constant space, O(n) = linear space, etc.
 *
 * Difficulty: Easy
 * Topics: Space Complexity, Memory Analysis
 */
public class ex2SpaceComplexity {

    /**
     * Snippet 1: Sum array elements
     */
    public static int snippet1(int[] arr) {
        int sum = 0;
        for (int num : arr) {
            sum += num;
        }
        return sum;
    }

    public static String getSpaceComplexitySnippet1() {
        // TODO: Return the space complexity of snippet1
        // Hint: How many extra variables do we use regardless of input size?
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Snippet 2: Create a copy of the array
     */
    public static int[] snippet2(int[] arr) {
        int[] copy = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            copy[i] = arr[i];
        }
        return copy;
    }

    public static String getSpaceComplexitySnippet2() {
        // TODO: Return the space complexity of snippet2
        // Hint: How does the size of 'copy' relate to the input size?
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Snippet 3: Recursive factorial
     */
    public static long snippet3(int n) {
        if (n <= 1) return 1;
        return n * snippet3(n - 1);
    }

    public static String getSpaceComplexitySnippet3() {
        // TODO: Return the space complexity of snippet3
        // Hint: Consider the call stack. How many recursive calls are made?
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Snippet 4: Create 2D array
     */
    public static int[][] snippet4(int n) {
        int[][] matrix = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                matrix[i][j] = i + j;
            }
        }
        return matrix;
    }

    public static String getSpaceComplexitySnippet4() {
        // TODO: Return the space complexity of snippet4
        // Hint: How many elements are in an n×n matrix?
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Snippet 5: Filter array into list
     */
    public static List<Integer> snippet5(int[] arr, int threshold) {
        List<Integer> result = new ArrayList<>();
        for (int num : arr) {
            if (num > threshold) {
                result.add(num);
            }
        }
        return result;
    }

    public static String getSpaceComplexitySnippet5() {
        // TODO: Return the space complexity of snippet5
        // Hint: In the worst case, how many elements might be in the result list?
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Snippet 6: Fibonacci with memoization
     */
    public static long snippet6(int n, Map<Integer, Long> memo) {
        if (n <= 1) return n;
        if (memo.containsKey(n)) return memo.get(n);

        long result = snippet6(n - 1, memo) + snippet6(n - 2, memo);
        memo.put(n, result);
        return result;
    }

    public static String getSpaceComplexitySnippet6() {
        // TODO: Return the space complexity of snippet6
        // Hint: Consider both the memo map size and the recursion stack depth
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public static void main(String[] args) {
        System.out.println("Big O Analysis - Exercise 2: Space Complexity");
        System.out.println("Implement getSpaceComplexity methods");

        try {
            System.out.println("Snippet 1 (sum array): " + getSpaceComplexitySnippet1());
            System.out.println("Snippet 2 (copy array): " + getSpaceComplexitySnippet2());
            System.out.println("Snippet 3 (recursive factorial): " + getSpaceComplexitySnippet3());
            System.out.println("Snippet 4 (2D array): " + getSpaceComplexitySnippet4());
            System.out.println("Snippet 5 (filter array): " + getSpaceComplexitySnippet5());
            System.out.println("Snippet 6 (fibonacci memoization): " + getSpaceComplexitySnippet6());
        } catch (UnsupportedOperationException e) {
            System.out.println("Not yet implemented!");
        }
    }
}
